core/src/main/scala/org/apache/spark/util/logging/DriverLogger.scala

```diff
@@ -0,0 +1,196 @@
+  }
+
+  def startSync(hadoopConf: Configuration): Unit = {
+    try {
+      // Setup a writer which moves the local file to hdfs continuously
+      val appId = Utils.sanitizeDirName(conf.getAppId)
+      writer = Some(new DfsAsyncWriter(appId, hadoopConf))
+    } catch {
+      case e: Exception =>
+        logError(s"Could not sync driver logs to spark dfs", e)
+    }
+  }
+
+  def stop(): Unit = {
+    try {
+      val fa = LogManager.getRootLogger.getAppender(DriverLogger.APPENDER_NAME)
+      LogManager.getRootLogger().removeAppender(DriverLogger.APPENDER_NAME)
+      Utils.tryLogNonFatalError(fa.close())
+      writer.foreach(_.closeWriter())
+    } catch {
+      case e: Exception =>
+        logError(s"Error in persisting driver logs", e)
+    } finally {
+      Utils.tryLogNonFatalError(JavaUtils.deleteRecursively(
+        FileUtils.getFile(localLogFile).getParentFile()))
+    }
+  }
+
+  // Visible for testing
+  private[spark] class DfsAsyncWriter(appId: String, hadoopConf: Configuration) extends Runnable
+    with Logging {
+
+    private var streamClosed = false
+    private val fileSystem: FileSystem = FileSystem.get(hadoopConf)
+    private val dfsLogFile: String = {
+      val rootDir = conf.get(DRIVER_LOG_DFS_DIR).get
+      if (!fileSystem.exists(new Path(rootDir))) {
+        throw new RuntimeException(s"${rootDir} does not exist." +
+          s" Please create this dir in order to sync driver logs")
+      }
+      FileUtils.getFile(rootDir, appId, DRIVER_LOG_FILE).getAbsolutePath()
+    }
+    private var inStream: InputStream = null
+    private var outputStream: FSDataOutputStream = null
+    try {
+      inStream = new BufferedInputStream(new FileInputStream(localLogFile))
+      outputStream = fileSystem.create(new Path(dfsLogFile), true)
+      fileSystem.setPermission(new Path(dfsLogFile), LOG_FILE_PERMISSIONS)
+    } catch {
+      case e: Exception =>
+        if (inStream != null) {
+          Utils.tryLogNonFatalError(inStream.close())
+        }
+        if (outputStream != null) {
+          Utils.tryLogNonFatalError(outputStream.close())
+        }
+        throw e
+    }
+    private val tmpBuffer = new Array[Byte](UPLOAD_CHUNK_SIZE)
+    private val threadpool = ThreadUtils.newDaemonSingleThreadScheduledExecutor("dfsSyncThread")
+    threadpool.scheduleWithFixedDelay(this, UPLOAD_INTERVAL_IN_SECS, UPLOAD_INTERVAL_IN_SECS,
+      TimeUnit.SECONDS)
+    logInfo(s"The local driver log file is being synced to spark dfs at: ${dfsLogFile}")
+
+    def run(): Unit = {
+      if (streamClosed) {
+        return
+      }
+      try {
+        var remaining = inStream.available()
+        while (remaining > 0) {
+          val read = inStream.read(tmpBuffer, 0, math.min(remaining, UPLOAD_CHUNK_SIZE))
+          outputStream.write(tmpBuffer, 0, read)
+          remaining -= read
+        }
+        outputStream.hflush()
+      } catch {
+        case e: Exception => logError("Failed to write to spark dfs", e)
```

Comment:
its not really "spark dfs".  Maybe "Failed copying driver logs to dfs".

If there is a consistent error, you're going to see this error msg a lot as you'll keep calling `run()`.  I'm not sure how common that will be and if we should do some special handling (eg. only include the full stack trace the first time or something like that).

Reply:
There can be cases when there are consistent error - such that someone deleted the file from filesystem, changed the permission, etc. But mostly these will be one off, so I am leaving the exception handling as it is for now. We can certainly change this later on if we discover otherwise.